### DISCUSSION TITLE
Changing the C-H kernel by SplitCHWRes and SplitCHParsed

### DISCUSSION THREAD

**Discussion by dongcongcong606** [idaholab/moose#18490] [Q&A Modules: Phase field] [answered]
Hi moose experts，
I want to add an anisotropic effect to mobility, _M_(c,angle)=_M_*_Function_(angle), where c is the order parameter and _angle_=atan*(gradc_y/gradc_x).  My  governing equation is "cTimeDerivative=grad _M_(c,angle)grad_w_  ; _w_ is an expression about the order parameter c".  So I wrote a related subroutine AnisotropicFirst.h/C to introduce _Function_(angle),  for example:

```
ANisotropicFirst::ANisotropicFirst(const InputParameters & Parameters)
  : Material(Parameters),
    // Get the parameters from the input file
    _delta(getParam<Real>("anisotropy_strength")),
    _j(getParam<unsigned int>("mode_number")),
    _theta0(getParam<Real>("reference_angle")),
    _mob(declareProperty<Real>("mob")),
    _dmob(declareProperty<Real>("dmob")),
    _dmobdgrad_op(declareProperty<RealGradient>("dmobdgrad_op")),
    _fun(declareProperty<Real>("fun")),
    _dfun(declareProperty<Real>("dfun")),
    _dfundgrad_op(declareProperty<RealGradient>("dfundgrad_op")),	
    _op(coupledValue("op")),
    _grad_op(coupledGradient("op"))	
{
  if(_mesh.dimension() != 2)
    mooseError("ANisotropicFirst requires a two-dimensional mesh.");
}

void
ANisotropicFirst::computeQpProperties()
{
  const Real tol = 1e-9;
  const Real cutoff = 1.0 - tol;
  // tagent of the gradient orientation angle
  Real n = 0.0;
  const Real nsq = _grad_op[_qp].norm_sq();
  if(nsq>tol)
    n = std::max(-cutoff, std::min((_grad_op[_qp](1) / _grad_op[_qp](0)) / std::sqrt(nsq), cutoff));

// calculate angle
  const Real angle = std::atan(n) * MathUtils::sign(_grad_op[_qp](1) / _grad_op[_qp](0));

// Compute derivative of angle wrt n
 const Real dangledn = MathUtils::sign((_grad_op[_qp](1)) / _grad_op[_qp](0)) / (1.0 + n * n);
 
   RealGradient dndgrad_op;
  if (nsq > tol)
  {
    dndgrad_op(0) = -2 * _grad_op[_qp](1) - _grad_op[_qp](1) * _grad_op[_qp](1) * _grad_op[_qp](1) / (_grad_op[_qp](0) * _grad_op[_qp](0));
    dndgrad_op(1) = _grad_op[_qp](0);
    dndgrad_op /= (_grad_op[_qp].norm_sq() * _grad_op[_qp].norm());
  }
 
  // Calculate  function of mobility

  _fun[_qp] = ((_delta * std::pow(cos(_j * (angle - _theta0 * libMesh::pi / 180)), 2.0) + 1.0) / (1.0 + _delta));
  _mob[_qp] = _mob_bar * _fun[_qp];
  
  // Calculate interfacial parameter function  's derivatives  
 _dfun[_qp] = -_delta * _j * std::sin(2.0 * _j * (angle - _theta0 * libMesh::pi / 180)) / (1.0 + _delta);
  _dmob[_qp] = _mob_bar * _dfun[_qp];
  
  // Compute derivatives of mobility and its derivative wrt grad_op
  _dfundgrad_op[_qp] = _dfun[_qp] * dangledn * dndgrad_op;
  _dmobdgrad_op[_qp] = _mob_bar * _dfundgrad_op[_qp];
}
```

And do I also need to introduce a coupling variable v("_Function_(angle)") into the kernel ? Is it enough to do it this way and include _v[_qp] or _grad_v[_qp]  in the residual and jacobian expressions? What am I missing? 

input file:
```
[./w_dot]
    variable = w
    type = CoupledTimeDerivative
    v = c
  [../]
  [./coupled_res]
    variable = w
    type = SplitCHWResF
    mob_name = M
  [../]
  [./coupled_parsed]
    variable = c
    type = SplitCHParsed
    f_name = F
    kappa_name = kappa_c
    w = w

  [./mobility]
    type = ParsedMaterial
    f_name = M
    args = c 
    material_property_names = 'kappa_c'
    function = 'if(c > -1.0 & c < 1.0,  (1.0 - c^2)^2, 0)'
    outputs = console
  [../]
  [./anisotropy_function]
   type = ANisotropicFirst
   op = c
   anisotropy_strength = 10
   mode_number = 1
   reference_angle = 90
 [../]
```

SplitCHWResBaseF_h_file:
```
private:
const VariableValue & _v;
_v(coupledValue("coupled_mobility"))
const MaterialProperty<Real> & _fun;
const MaterialProperty<Real> & _dfun;
const MaterialProperty<RealGradient> & _dfundgrad_op;
```

SplitCHWResBaseF_C_file:

```
_fun(getMaterialProperty<Real>("fun_name")),
_dfun(getMaterialProperty<Real>("dfun_name")),
_dfundgrad_op(getMaterialProperty<RealGradient>("dfundgrad_op_name")),
```

Residual:

```
Real
SplitCHWResBaseF::computeQpResidual()
{
  return _v[_qp] * _mob[_qp] * _grad_w[_qp] * _grad_test[_i][_qp];
}
```

Jacobian:

```
Real
SplitCHWResBaseF::computeQpResidual()
{
  return _v[_qp] * _mob[_qp] * _grad_w[_qp] * _grad_test[_i][_qp];
}
```

Off-Jacobian:

```
Real
SplitCHWResBase4::computeQpOffDiagJacobian(unsigned int jvar)
{
  if (_w_var == jvar)
    return computeQpWJacobian();
const unsigned int cvar = mapJvarToCvar(jvar);

const RealGradient a(-_grad_op[_qp](1), _grad_op[_qp](0), 0);

 return a * _dfundgrad_op[_qp] * (*_dmobdarg[cvar])[_qp] * _phi[_j][_qp] * _grad_w[_qp] * _grad_test[_i][_qp];
}
```

I am puzzled, and I 'd like to accept your advice and guidance. May someone help me? Thank you very much.

**Comment by dschwen** [comment, marked as answer]
First of all 

```
angle=std::atan2(gradc_y, gradc_x);
```

for a well defined angle!

Next, you are mixing coupled variables (like _v_) and material properties (like `dfundgrad_op`). What Kernel computes _v_? As far as I can tell you need to compute a mobility property and it's derivatives w.r.t. all coupled variables. The mobility does not appear in the variational derivative in the CH equation, so its derivative w.r.t. to order parameter gradients should not matter, right?

I also suggest you utilize the `DerivativeMaterialInterface` to declare and get derivative material properties for better interoperability with the rest of the phase field module.

In short, this should be really simple to implement with _only_ a material for computing the mobility and simply using all stock _split_ CH formulation kernels.

**Reply by dongcongcong606** [reply]
First of all，thanks for your advice, I'll learn "DerivativeMaterialInterface".
Secondly, I am a little confused why Angle is defined as "angle=std::atan2(gradc_y, gradc_x);" instead of "std::atan(n) * MathUtils::sign(_grad_op_qp / _grad_op_qp);" . Could you tell me the difference ? (Silly question though it may be. )

**Reply by GiudGiud** [reply]
https://stackoverflow.com/questions/283406/what-is-the-difference-between-atan-and-atan2-in-c

**Reply by dschwen** [reply]
If you have numerator and denominator of the tangent available _always_ just use `std::atan2`.
